src/test/java/com/microsoft/sqlserver/jdbc/requestboundary/RequestBoundaryMethodsTest.java

```diff
@@ -0,0 +1,421 @@
+
+                cs = con.prepareCall("{call sp_server_info}");
+                cs.execute();
+                con.endRequest();
+
+                assertTrue(stmt.isClosed());
+                assertTrue(ps.isClosed());
+                assertTrue(stmt1.isClosed());
+                assertTrue(cs.isClosed());
+                assertTrue(rs1.isClosed());
+            }
+        }
+        finally {
+            if (null != stmt) {
+                stmt.close();
+            }
+            if (null != stmt1) {
+                stmt1.close();
+            }
+            if (null != ps) {
+                ps.close();
+            }
+            if (null != cs) {
+                cs.close();
+            }
+            if (null != con) {
+                con.close();
+            }
+        }
+    }
+
+    /**
+     * Tests Request Boundary methods in a multi-threaded environment.
+     * 
+     * @throws SQLException
+     */
+    @Test
+    public void testThreads() throws SQLException {
+        try {
+            con = connect();
+            if (Utils.isJDBC43AndGreater(con)) {
+                Thread thread1 = new Thread() {
+                    public void run() {
+                        try {
+                            con.setNetworkTimeout(null, 100);
+                            con.setHoldability(ResultSet.CLOSE_CURSORS_AT_COMMIT);
+                        }
+                        catch (SQLException e) {
+                            e.printStackTrace();
+                        }
+                    }
+                };
+
+                Thread thread2 = new Thread() {
+                    public void run() {
+                        try {
+                            stmt = con.createStatement();
+                            ResultSet rs = stmt.executeQuery("SELECT 1");
+                            rs.next();
+                            assertEquals(1, rs.getInt(1));
+                        }
+                        catch (SQLException e) {
+                            e.printStackTrace();
+                        }
+                    }
+                };
+
+                Thread thread3 = new Thread() {
+                    public void run() {
+                        try {
+                            pstmt = con.prepareStatement("SELECT 1");
+                            ResultSet rs = pstmt.executeQuery();
+                            rs.next();
+                            assertEquals(1, rs.getInt(1));
+                        }
+                        catch (SQLException e) {
+                            e.printStackTrace();
+                        }
+
+                    }
+                };
+
+                int originalNetworkTimeout = con.getNetworkTimeout();
+                int originalHoldability = con.getHoldability();
+                con.beginRequest();
+                thread1.start();
+                thread2.start();
+                thread3.start();

```

Comment:
You could use `CountDownLatch` for triggering all threads at exact same time, I guess that is what we should be testing here.

Reply:
No need to trigger threads at the same time, that's not the point of the test. We just need to make sure connection is back to its original state once the threads finished execution.